hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -8,18 +8,38 @@ object InferPType {
   def apply(ir: IR, env: Env[PType]): Unit = {
     assert(ir._pType2 == null)
     ir._pType2 = ir match {
-      case I32(_) => PInt32()
-      case I64(_) => PInt64()
-      case F32(_) => PFloat32()
-      case F64(_) => PFloat64()
-      case Str(_) => PString()
-      case Literal(t, _) => PType.canonical(t)
-      case True() | False() => PBoolean()
+        // TODO: can these all be given null as a value, like Literals?
+      case I32(_) => PInt32(true)
+      case I64(_) => PInt64(true)
+      case F32(_) => PFloat32(true)
+      case F64(_) => PFloat64(true)
+      case Str(_) => PString(true)
+      // TODO: Do we want this to be required? We allow coerce from NA wrapped Literal
```

Comment:
I think it's fine for this to be required, as long as we can enforce that `v` is never null.

(actually, if you wanted to be extra sure, this could be: `case Literal(t, v) => PType.canonical(t, v != null)`)

Reply:
is safe:
```
final case class Literal(_typ: Type, value: Annotation) extends IR {
  require(!CanEmit(_typ))
  require(value != null)
}
```

Reply:
Should we do the same for the primitive nodes?

Reply:
technically, a Scala primitive type is already not allowed to be null.

Reply:
Make sense, thanks